docs/config-schema.md
<!--- GENERATED BY gomplate from scripts/docs/templates/top-config.md.tmpl --->
# Agent Configuration

The agent is configured primarily by a YAML document conventionally located at
`/etc/signalfx/agent.yaml`.  The location of the config file can be specified
by the `-config` flag to the agent binary (`signalfx-agent`).

## Configuring your endpoint

By default, the Smart Agent will send data to the `us0` realm.
If you are not in this realm, you will need to explicitly set the
`signalFxRealm` option in your config like this:

```
signalFxRealm: YOUR_SIGNALFX_REALM
```

To determine if you are in a different realm and need to
explicitly set the endpoints, check your profile page in the SignalFx
web application.

If you want to explicitly set the ingest, API server, and trace endpoint URLs,
you can set them individually like so:

```
ingestUrl: "https://ingest.YOUR_SIGNALFX_REALM.signalfx.com"
apiUrl: "https://api.YOUR_SIGNALFX_REALM.signalfx.com"
traceEndpointUrl: "https://ingest.YOUR_SIGNALFX_REALM.signalfx.com/v1/trace"
```

They will default to the endpoints for the realm configured in `signalFxRealm`
if not set.


## Config Schema


| Config option | Required | Type | Description |
| --- | --- | --- | --- |
| `signalFxAccessToken` | no | string | The access token for the org that should receive the metrics emitted by the agent. |
| `ingestUrl` | no | string | The URL of SignalFx ingest server.  Should be overridden if using the SignalFx Gateway.  If not set, this will be determined by the `signalFxRealm` option below.  If you want to send trace spans to a different location, set the `traceEndpointUrl` option.  If you want to send events to a different location, set the `eventEndpointUrl` option. |
| `eventEndpointUrl` | no | string | The full URL (including path) to the event ingest server.  If this is not set, all events will be sent to the same place as `ingestUrl` above. |
| `traceEndpointUrl` | no | string | The full URL (including path) to the trace ingest server.  If this is not set, all trace spans will be sent to the same place as `ingestUrl` above. |
| `apiUrl` | no | string | The SignalFx API base URL.  If not set, this will determined by the `signalFxRealm` option below. |
| `signalFxRealm` | no | string | The SignalFx Realm that the organization you want to send to is a part of.  This defaults to the original realm (`us0`) but if you are setting up the agent for the first time, you quite likely need to change this. (**default:** `"us0"`) |
| `hostname` | no | string | The hostname that will be reported as the `host` dimension. If blank, this will be auto-determined by the agent based on a reverse lookup of the machine's IP address. |
| `useFullyQualifiedHost` | no | bool | If true (the default), and the `hostname` option is not set, the hostname will be determined by doing a reverse DNS query on the IP address that is returned by querying for the bare hostname.  This is useful in cases where the hostname reported by the kernel is a short name. (**default**: `true`) |
| `disableHostDimensions` | no | bool | Our standard agent model is to collect metrics for services running on the same host as the agent.  Therefore, host-specific dimensions (e.g. `host`, `AWSUniqueId`, etc) are automatically added to every datapoint that is emitted from the agent by default.  Set this to true if you are using the agent primarily to monitor things on other hosts.  You can set this option at the monitor level as well. (**default:** `false`) |
| `intervalSeconds` | no | integer | How often to send metrics to SignalFx.  Monitors can override this individually. (**default:** `10`) |
| `globalDimensions` | no | map of strings | Dimensions (key:value pairs) that will be added to every datapoint emitted by the agent. To specify that all metrics should be high-resolution, add the dimension `sf_hires: 1` |
| `environment` | no | string | The logical environment/cluster that this agent instance is running in. All of the services that this instance monitors should be in the same environment as well. This value, if provided, will be synced as a property onto the `host` dimension, or onto any cloud-provided specific dimensions (`AWSUniqueId`, `gcp_id`, and `azure_resource_id`) when available. Example values: "prod-usa", "dev" |
| `cluster` | no | string | This configuration is deprecated and has been replaced with `environment`. Setting `environment` will override `cluster`.  If `environment` is not set then `cluster` will be used as a fallback. |
| `syncClusterOnHostDimension` | no | bool | If true, force syncing of the `cluster` property on the `host` dimension, even when cloud-specific dimensions are present. (**default:** `false`) |
| `validateDiscoveryRules` | no | bool | If true, a warning will be emitted if a discovery rule contains variables that will never possibly match a rule.  If using multiple observers, it is convenient to set this to false to suppress spurious errors. (**default:** `true`) |
| `observers` | no | [list of objects (see below)](#observers) | A list of observers to use (see observer config) |
| `monitors` | no | [list of objects (see below)](#monitors) | A list of monitors to use (see monitor config) |
| `writer` | no | [object (see below)](#writer) | Configuration of the datapoint/event writer |
| `logging` | no | [object (see below)](#logging) | Log configuration |
| `collectd` | no | [object (see below)](#collectd) | Configuration of the managed collectd subprocess |
| `enableBuiltInFiltering` | no | bool | This must be unset or explicitly set to true. In prior versions of the agent, there was a filtering mechanism that relied heavily on an external whitelist.json file to determine which metrics were sent by default.  This is all inherent to the agent now and the old style of filtering is no longer available. (**default:** `true`) |
| `metricsToInclude` | no | [list of objects (see below)](#metricstoinclude) | A list of metric filters that will whitelist/include metrics.  These filters take priority over the filters specified in `metricsToExclude`. |
| `metricsToExclude` | no | [list of objects (see below)](#metricstoexclude) | A list of metric filters |
| `propertiesToExclude` | no | [list of objects (see below)](#propertiestoexclude) | A list of properties filters |
| `internalStatusHost` | no | string | The host on which the internal status server will listen.  The internal status HTTP server serves internal metrics and diagnostic information about the agent and can be scraped by the `internal-metrics` monitor. Can be set to `0.0.0.0` if you want to monitor the agent from another host.  If you set this to blank/null, the internal status server will not be started.  See `internalStatusPort`. (**default:** `"localhost"`) |
| `internalStatusPort` | no | integer | The port on which the internal status server will listen.  See `internalStatusHost`. (**default:** `8095`) |
| `profiling` | no | bool | Enables Go pprof endpoint on port 6060 that serves profiling data for development (**default:** `false`) |
| `profilingHost` | no | string | The host/ip address for the pprof profile server to listen on. `profiling` must be enabled for this to have any effect. (**default:** `"127.0.0.1"`) |
| `profilingPort` | no | integer | The port for the pprof profile server to listen on. `profiling` must be enabled for this to have any effect. (**default:** `6060`) |
| `bundleDir` | no | string | Path to the directory holding the agent dependencies.  This will normally be derived automatically. Overrides the envvar SIGNALFX_BUNDLE_DIR if set. |
| `scratch` | no | any | This exists purely to give the user a place to put common yaml values to reference in other parts of the config file. |
| `configSources` | no | [object (see below)](#configsources) | Configuration of remote config stores |
| `procPath` | no | string | Path to the host's `/proc` filesystem. This is useful for containerized environments. (**default:** `"/proc"`) |
| `etcPath` | no | string | Path to the host's `/etc` directory. This is useful for containerized environments. (**default:** `"/etc"`) |
| `varPath` | no | string | Path to the host's `/var` directory. This is useful for containerized environments. (**default:** `"/var"`) |
| `runPath` | no | string | Path to the host's `/run` directory. This is useful for containerized environments. (**default:** `"/run"`) |
| `sysPath` | no | string | Path to the host's `/sys` directory. This is useful for containerized environments. (**default:** `"/sys"`) |


## observers
The **nested** `observers` config object has the following fields:

The following are generic options that apply to all observers.  Each observer type has its own set of additional configuration options, detailed in [Observer Config](./observer-config.md).

| Config option | Required | Type | Description |
| --- | --- | --- | --- |
| `type` | no | string | The type of the observer |



## monitors
The **nested** `monitors` config object has the following fields:

The following are generic options that apply to all monitors.  Each monitor type has its own set of additional configuration options, detailed in [Monitor Config](./monitor-config.md).

| Config option | Required | Type | Description |
| --- | --- | --- | --- |
| `type` | no | string | The type of the monitor |
| `discoveryRule` | no | string | The rule used to match up this configuration with a discovered endpoint. If blank, the configuration will be run immediately when the agent is started.  If multiple endpoints match this rule, multiple instances of the monitor type will be created with the same configuration (except different host/port). |
| `validateDiscoveryRule` | no | bool | If true, a warning will be emitted if a discovery rule contains variables that will never possibly match a rule.  If using multiple observers, it is convenient to set this to false to suppress spurious errors.  The top-level setting `validateDiscoveryRules` acts as a default if this isn't set. (**default:** `"false"`) |
| `extraDimensions` | no | map of strings | A set of extra dimensions (key:value pairs) to include on datapoints emitted by the monitor(s) created from this configuration. To specify metrics from this monitor should be high-resolution, add the dimension `sf_hires: 1` |
| `extraDimensionsFromEndpoint` | no | map of strings | A mapping of extra dimension names to a [discovery rule expression](https://docs.signalfx.com/en/latest/integrations/agent/auto-discovery.html) that is used to derive the value of the dimension.  For example, to use a certain container label as a dimension, you could use something like this in your monitor config block: `extraDimensionsFromEndpoint: {env: 'Get(container_labels, "myapp.com/environment")'}` |
| `configEndpointMappings` | no | map of strings | A set of mappings from a configuration option on this monitor to attributes of a discovered endpoint.  The keys are the config option on this monitor and the value can be any valid expression used in discovery rules. |
| `intervalSeconds` | no | integer | The interval (in seconds) at which to emit datapoints from the monitor(s) created by this configuration.  If not set (or set to 0), the global agent intervalSeconds config option will be used instead. (**default:** `0`) |
| `solo` | no | bool | If one or more configurations have this set to true, only those configurations will be considered. This setting can be useful for testing. (**default:** `false`) |
| `datapointsToExclude` | no | [list of objects (see below)](#datapointstoexclude) | A list of datapoint filters.  These filters allow you to comprehensively define which datapoints to exclude by metric name or dimension set, as well as the ability to define overrides to re-include metrics excluded by previous patterns within the same filter item.  See [monitor filtering](./filtering.md#additional-monitor-level-filtering) for examples and more information. |
| `disableHostDimensions` | no | bool | Some monitors pull metrics from services not running on the same host and should not get the host-specific dimensions set on them (e.g. `host`, `AWSUniqueId`, etc).  Setting this to `true` causes those dimensions to be omitted.  You can disable this globally with the `disableHostDimensions` option on the top level of the config. (**default:** `false`) |
| `disableEndpointDimensions` | no | bool | This can be set to true if you don't want to include the dimensions that are specific to the endpoint that was discovered by an observer.  This is useful when you have an endpoint whose identity is not particularly important since it acts largely as a proxy or adapter for other metrics. (**default:** `false`) |
| `dimensionTransformations` | no | map of strings | A map from dimension names emitted by the monitor to the desired dimension name that will be emitted in the datapoint that goes to SignalFx.  This can be useful if you have custom metrics from your applications and want to make the dimensions from a monitor match those. Also can be useful when scraping free-form metrics, say with the `prometheus-exporter` monitor.  Right now, only static key/value transformations are supported.  Note that filtering by dimensions will be done on the *original* dimension name and not the new name. Note that it is possible to remove unwanted dimensions via this configuration, by making the desired dimension name an empty string. |
| `extraMetrics` | no | list of strings | Extra metrics to enable besides the default included ones.  This is an [overridable filter](https://docs.signalfx.com/en/latest/integrations/agent/filtering.html#overridable-filtering). |
| `extraGroups` | no | list of strings | Extra metric groups to enable in addition to the metrics that are emitted by default.  A metric group is simply a collection of metrics, and they are defined in each monitor's documentation. |


## datapointsToExclude
The **nested** `datapointsToExclude` config object has the following fields:



| Config option | Required | Type | Description |
| --- | --- | --- | --- |
| `dimensions` | no | map of any | A map of dimension key/values to match against.  All key/values must match a datapoint for it to be matched.  The map values can be either a single string or a list of strings. |
| `metricNames` | no | list of strings | A list of metric names to match against |
| `metricName` | no | string | A single metric name to match against |
| `monitorType` | no | string | (**Only applicable for the top level filters**) Limits this scope of the filter to datapoints from a specific monitor. If specified, any datapoints not from this monitor type will never match against this filter. |
| `negated` | no | bool | (**Only applicable for the top level filters**) Negates the result of the match so that it matches all datapoints that do NOT match the metric name and dimension values given. This does not negate monitorType, if given. (**default:** `false`) |




## writer
The **nested** `writer` config object has the following fields:



| Config option | Required | Type | Description |
| --- | --- | --- | --- |
| `datapointMaxBatchSize` | no | integer | The maximum number of datapoints to include in a batch before sending the batch to the ingest server.  Smaller batch sizes than this will be sent if datapoints originate in smaller chunks. (**default:** `1000`) |
| `maxDatapointsBuffered` | no | integer | The maximum number of datapoints that are allowed to be buffered in the agent (i.e. received from a monitor but have not yet received confirmation of successful receipt by the target ingest/gateway server downstream).  Any datapoints that come in beyond this number will overwrite existing datapoints if they have not been sent yet, starting with the oldest. (**default:** `25000`) |
| `traceSpanMaxBatchSize` | no | integer | The analogue of `datapointMaxBatchSize` for trace spans. (**default:** `1000`) |

pkg/core/config/config.go
// Package config contains configuration structures and related helper logic for all
// agent components.
package config

import (
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	set "gopkg.in/fatih/set.v0"

	"github.com/mitchellh/hashstructure"
	"github.com/pkg/errors"
	"github.com/signalfx/signalfx-agent/pkg/core/common/constants"
	"github.com/signalfx/signalfx-agent/pkg/core/config/sources"
	"github.com/signalfx/signalfx-agent/pkg/core/config/validation"
	"github.com/signalfx/signalfx-agent/pkg/utils"
	"github.com/signalfx/signalfx-agent/pkg/utils/hostfs"
	log "github.com/sirupsen/logrus"
)

const (
	TraceExportFormatSAPM   = "sapm"
	TraceExportFormatZipkin = "zipkin"
)

// Config is the top level config struct for configurations that are common to all platoforms
type Config struct {
	// The access token for the org that should receive the metrics emitted by
	// the agent.
	SignalFxAccessToken string `yaml:"signalFxAccessToken" neverLog:"true"`
	// The URL of SignalFx ingest server.  Should be overridden if using the
	// SignalFx Gateway.  If not set, this will be determined by the
	// `signalFxRealm` option below.  If you want to send trace spans to a
	// different location, set the `traceEndpointUrl` option.  If you want to
	// send events to a different location, set the `eventEndpointUrl` option.
	IngestURL string `yaml:"ingestUrl"`
	// The full URL (including path) to the event ingest server.  If this is
	// not set, all events will be sent to the same place as `ingestUrl`
	// above.
	EventEndpointURL string `yaml:"eventEndpointUrl"`
	// The full URL (including path) to the trace ingest server.  If this is
	// not set, all trace spans will be sent to the same place as `ingestUrl`
	// above.
	TraceEndpointURL string `yaml:"traceEndpointUrl"`
	// The SignalFx API base URL.  If not set, this will determined by the
	// `signalFxRealm` option below.
	APIURL string `yaml:"apiUrl"`
	// The SignalFx Realm that the organization you want to send to is a part
	// of.  This defaults to the original realm (`us0`) but if you are setting
	// up the agent for the first time, you quite likely need to change this.
	SignalFxRealm string `yaml:"signalFxRealm" default:"us0"`
	// The hostname that will be reported as the `host` dimension. If blank,
	// this will be auto-determined by the agent based on a reverse lookup of
	// the machine's IP address.
	Hostname string `yaml:"hostname"`
	// If true (the default), and the `hostname` option is not set, the
	// hostname will be determined by doing a reverse DNS query on the IP
	// address that is returned by querying for the bare hostname.  This is
	// useful in cases where the hostname reported by the kernel is a short
	// name. (**default**: `true`)
	UseFullyQualifiedHost *bool `yaml:"useFullyQualifiedHost" noDefault:"true"`
	// Our standard agent model is to collect metrics for services running on
	// the same host as the agent.  Therefore, host-specific dimensions (e.g.
	// `host`, `AWSUniqueId`, etc) are automatically added to every datapoint
	// that is emitted from the agent by default.  Set this to true if you are
	// using the agent primarily to monitor things on other hosts.  You can set
	// this option at the monitor level as well.
	DisableHostDimensions bool `yaml:"disableHostDimensions" default:"false"`
	// How often to send metrics to SignalFx.  Monitors can override this
	// individually.
	IntervalSeconds int `yaml:"intervalSeconds" default:"10"`
	// Dimensions (key:value pairs) that will be added to every datapoint emitted by the agent.
	// To specify that all metrics should be high-resolution, add the dimension `sf_hires: 1`
	GlobalDimensions map[string]string `yaml:"globalDimensions" default:"{}"`
	// The logical environment/cluster that this agent instance is running in.
	// All of the services that this instance monitors should be in the same
	// environment as well. This value, if provided, will be synced as a
	// property onto the `host` dimension, or onto any cloud-provided specific
	// dimensions (`AWSUniqueId`, `gcp_id`, and `azure_resource_id`) when
	// available. Example values: "prod-usa", "dev"
	Environment string `yaml:"environment"`
	// This configuration is deprecated and has been replaced with `environment`.
	// Setting `environment` will override `cluster`.  If `environment` is not set
	// then `cluster` will be used as a fallback.
	Cluster string `yaml:"cluster"`
	// If true, force syncing of the `cluster` property on the `host` dimension,
	// even when cloud-specific dimensions are present.
	SyncClusterOnHostDimension bool `yaml:"syncClusterOnHostDimension"`
	// If true, a warning will be emitted if a discovery rule contains
	// variables that will never possibly match a rule.  If using multiple
	// observers, it is convenient to set this to false to suppress spurious
	// errors.
	ValidateDiscoveryRules *bool `yaml:"validateDiscoveryRules" default:"true"`
	// A list of observers to use (see observer config)
	Observers []ObserverConfig `yaml:"observers" default:"[]"`
	// A list of monitors to use (see monitor config)
	Monitors []MonitorConfig `yaml:"monitors" default:"[]"`
	// Configuration of the datapoint/event writer
	Writer WriterConfig `yaml:"writer"`
	// Log configuration
	Logging LogConfig `yaml:"logging" default:"{}"`
	// Configuration of the managed collectd subprocess
	Collectd CollectdConfig `yaml:"collectd" default:"{}"`
	// This must be unset or explicitly set to true. In prior versions of the
	// agent, there was a filtering mechanism that relied heavily on an
	// external whitelist.json file to determine which metrics were sent by
	// default.  This is all inherent to the agent now and the old style of
	// filtering is no longer available.
	EnableBuiltInFiltering *bool `yaml:"enableBuiltInFiltering" default:"true"`
	// A list of metric filters that will whitelist/include metrics.  These
	// filters take priority over the filters specified in `metricsToExclude`.
	MetricsToInclude []MetricFilter `yaml:"metricsToInclude" default:"[]"`
	// A list of metric filters
	MetricsToExclude []MetricFilter `yaml:"metricsToExclude" default:"[]"`
	// A list of properties filters
	PropertiesToExclude []PropertyFilterConfig `yaml:"propertiesToExclude" default:"[]"`

	// The host on which the internal status server will listen.  The internal
	// status HTTP server serves internal metrics and diagnostic information
	// about the agent and can be scraped by the `internal-metrics` monitor.
	// Can be set to `0.0.0.0` if you want to monitor the agent from another
	// host.  If you set this to blank/null, the internal status server will
	// not be started.  See `internalStatusPort`.
	InternalStatusHost string `yaml:"internalStatusHost" default:"localhost"`
	// The port on which the internal status server will listen.  See
	// `internalStatusHost`.
	InternalStatusPort uint16 `yaml:"internalStatusPort" default:"8095"`

	// Enables Go pprof endpoint on port 6060 that serves profiling data for
	// development
	EnableProfiling bool `yaml:"profiling" default:"false"`
	// The host/ip address for the pprof profile server to listen on.
	// `profiling` must be enabled for this to have any effect.
	ProfilingHost string `yaml:"profilingHost" default:"127.0.0.1"`
	// The port for the pprof profile server to listen on. `profiling` must be
	// enabled for this to have any effect.
	ProfilingPort int `yaml:"profilingPort" default:"6060"`
	// Path to the directory holding the agent dependencies.  This will
	// normally be derived automatically. Overrides the envvar
	// SIGNALFX_BUNDLE_DIR if set.
	BundleDir string `yaml:"bundleDir"`
	// This exists purely to give the user a place to put common yaml values to
	// reference in other parts of the config file.
	Scratch interface{} `yaml:"scratch" neverLog:"omit"`
	// Configuration of remote config stores
	Sources sources.SourceConfig `yaml:"configSources"`
	// Path to the host's `/proc` filesystem.
	// This is useful for containerized environments.
	ProcPath string `yaml:"procPath" default:"/proc"`
	// Path to the host's `/etc` directory.
	// This is useful for containerized environments.
	EtcPath string `yaml:"etcPath" default:"/etc"`
	// Path to the host's `/var` directory.
	// This is useful for containerized environments.
	VarPath string `yaml:"varPath" default:"/var"`
	// Path to the host's `/run` directory.
	// This is useful for containerized environments.
	RunPath string `yaml:"runPath" default:"/run"`
	// Path to the host's `/sys` directory.
	// This is useful for containerized environments.
	SysPath string `yaml:"sysPath" default:"/sys"`
}

// remapValues is used to move deprecated config values to newer keys.
// a warning should be logged for each migrated value
func (c *Config) remapValues() {
	// if environment is empty, then set it to conf.cluster for legacy behavior
	if c.Environment == "" && c.Cluster != "" {
		log.WithFields(log.Fields{
			"cluster":     c.Cluster,
			"environment": c.Environment,
		}).Warn("Config value 'cluster' is deprecated, please use `environment` instead")
		c.Environment = c.Cluster
	}
}

func (c *Config) initialize() (*Config, error) {
	if c.SignalFxRealm != "" {
		if c.IngestURL == "" {
			c.IngestURL = fmt.Sprintf("https://ingest.%s.signalfx.com", c.SignalFxRealm)
		}
		if c.APIURL == "" {
			c.APIURL = fmt.Sprintf("https://api.%s.signalfx.com", c.SignalFxRealm)
		}
	}

	c.Writer.initialize()
	c.setupEnvironment()

	if err := c.validate(); err != nil {
		return nil, errors.WithMessage(err, "configuration is invalid")
	}

	c.remapValues()
	err := c.propagateValuesDown()
	if err != nil {
		return nil, err
	}

	return c, nil
}

// Setup envvars that will be used by collectd to use the bundled dependencies
// instead of looking to the normal system paths.
func (c *Config) setupEnvironment() {
	if c.BundleDir == "" {
		c.BundleDir = os.Getenv(constants.BundleDirEnvVar)
	}
	if c.BundleDir == "" {
		exePath, err := os.Executable()
		if err != nil {
			panic("Cannot determine agent executable path, cannot continue")
		}
		c.BundleDir, err = filepath.Abs(filepath.Join(filepath.Dir(exePath), ".."))
		if err != nil {
			panic("Cannot determine absolute path of executable parent dir " + exePath)
		}
	}
	os.Setenv(constants.BundleDirEnvVar, c.BundleDir)

	os.Setenv("JAVA_HOME", filepath.Join(c.BundleDir, "jre"))
	// set the environment variables for gopsutil based on configured values
	os.Setenv(hostfs.HostProcVar, c.ProcPath)
	os.Setenv(hostfs.HostEtcVar, c.EtcPath)
	os.Setenv(hostfs.HostVarVar, c.VarPath)
	os.Setenv(hostfs.HostRunVar, c.RunPath)
	os.Setenv(hostfs.HostSysVar, c.SysPath)
}

// Validate everything that we can about the main config
func (c *Config) validate() error {
	if err := validation.ValidateStruct(c); err != nil {
		return err
	}

	if !*c.EnableBuiltInFiltering {
		return errors.New("enableBuiltInFiltering must be true or unset, false is no longer supported")
	}

	if _, err := url.Parse(c.IngestURL); err != nil {
		return fmt.Errorf("%s is not a valid ingest URL: %v", c.IngestURL, err)
	}

	if _, err := url.Parse(c.APIURL); err != nil {
		return fmt.Errorf("%s is not a valid API URL: %v", c.APIURL, err)
	}

	if _, err := url.Parse(c.EventEndpointURL); err != nil {
		return fmt.Errorf("%s is not a valid event endpoint URL: %v", c.EventEndpointURL, err)
	}

	if c.TraceEndpointURL != "" {
		if _, err := url.Parse(c.TraceEndpointURL); err != nil {
			return fmt.Errorf("%s is not a valid trace endpoint URL: %v", c.TraceEndpointURL, err)
		}
	}

	return c.Collectd.Validate()
}

// Send values from the top of the config down to nested configs that might
// need them
func (c *Config) propagateValuesDown() error {
	for i := range c.Monitors {
		if err := c.Monitors[i].Validate(); err != nil {
			return fmt.Errorf("monitor config for type '%s' is invalid: %v", c.Monitors[i].Type, err)
		}
		if c.Monitors[i].ValidateDiscoveryRule == nil {
			c.Monitors[i].ValidateDiscoveryRule = c.ValidateDiscoveryRules
		}
		if c.Monitors[i].ProcPath == "" {
			c.Monitors[i].ProcPath = c.ProcPath
		}
	}

	anyCollectdMonitors := false
	for i := range c.Monitors {
		anyCollectdMonitors = anyCollectdMonitors || c.Monitors[i].IsCollectdBased()
	}

	c.Collectd.DisableCollectd = c.Collectd.DisableCollectd || !anyCollectdMonitors
	c.Collectd.IntervalSeconds = utils.FirstNonZero(c.Collectd.IntervalSeconds, c.IntervalSeconds)
	c.Collectd.BundleDir = c.BundleDir

	c.Writer.MetricsToInclude = c.MetricsToInclude
	c.Writer.MetricsToExclude = c.MetricsToExclude
	c.Writer.IngestURL = c.IngestURL
	c.Writer.APIURL = c.APIURL
	c.Writer.EventEndpointURL = c.EventEndpointURL
	c.Writer.TraceEndpointURL = c.TraceEndpointURL
	c.Writer.SignalFxAccessToken = c.SignalFxAccessToken
	c.Writer.GlobalDimensions = c.GlobalDimensions
	c.Writer.Environment = c.Environment

	return nil
}

// CustomConfigurable should be implemented by config structs that have the
// concept of generic other config that is initially deserialized into a
// map[string]interface{} to be later transformed to another form.
type CustomConfigurable interface {
	ExtraConfig() (map[string]interface{}, error)
}

// LogConfig contains configuration related to logging
type LogConfig struct {
	// Valid levels include `debug`, `info`, `warn`, `error`.  Note that
	// `debug` logging may leak sensitive configuration (e.g. passwords) to the
	// agent output.
	Level string `yaml:"level" default:"info"`
	// The log output format to use.  Valid values are: `text`, `json`.
	Format string `yaml:"format" validate:"oneof=text json" default:"text"`
	// TODO: Support log file output and other log targets
}

// LogrusLevel returns a logrus log level based on the configured level in
// LogConfig.
func (lc *LogConfig) LogrusLevel() *log.Level {
	if lc.Level != "" {
		level, err := log.ParseLevel(lc.Level)
		if err != nil {
			log.WithFields(log.Fields{
				"level": lc.Level,
			}).Error("Invalid log level")
			return nil
		}
		return &level
	}
	return nil
}

// LogrusFormatter returns the formatter to use based on the config
func (lc *LogConfig) LogrusFormatter() log.Formatter {
	switch lc.Format {
	case "json":
		return &log.JSONFormatter{}
	default:
		return &log.TextFormatter{}
	}
}

var validCollectdLogLevels = set.NewNonTS("debug", "info", "notice", "warning", "err")

// CollectdConfig high-level configurations
type CollectdConfig struct {
	// If you won't be using any collectd monitors, this can be set to true to
	// prevent collectd from pre-initializing
	DisableCollectd bool `yaml:"disableCollectd" default:"false"`
	// How many read intervals before abandoning a metric. Doesn't affect much
	// in normal usage.
	// See [Timeout](https://collectd.org/documentation/manpages/collectd.conf.5.shtml#timeout_iterations).
	Timeout int `yaml:"timeout" default:"40"`
	// Number of threads dedicated to executing read callbacks. See
	// [ReadThreads](https://collectd.org/documentation/manpages/collectd.conf.5.shtml#readthreads_num)
	ReadThreads int `yaml:"readThreads" default:"5"`
	// Number of threads dedicated to writing value lists to write callbacks.
	// This should be much less than readThreads because writing is batched in
	// the write_http plugin that writes back to the agent.
	// See [WriteThreads](https://collectd.org/documentation/manpages/collectd.conf.5.shtml#writethreads_num).
	WriteThreads int `yaml:"writeThreads" default:"2"`
	// The maximum numbers of values in the queue to be written back to the
	// agent from collectd.  Since the values are written to a local socket
	// that the agent exposes, there should be almost no queuing and the
	// default should be more than sufficient. See
	// [WriteQueueLimitHigh](https://collectd.org/documentation/manpages/collectd.conf.5.shtml#writequeuelimithigh_highnum)
	WriteQueueLimitHigh int `yaml:"writeQueueLimitHigh" default:"500000"`
	// The lowest number of values in the collectd queue before which metrics
	// begin being randomly dropped.  See
	// [WriteQueueLimitLow](https://collectd.org/documentation/manpages/collectd.conf.5.shtml#writequeuelimitlow_lownum)
	WriteQueueLimitLow int `yaml:"writeQueueLimitLow" default:"400000"`
	// Collectd's log level -- info, notice, warning, or err
	LogLevel string `yaml:"logLevel" default:"notice"`
	// A default read interval for collectd plugins.  If zero or undefined,
	// will default to the global agent interval.  Some collectd python
	// monitors do not support overridding the interval at the monitor level,
	// but this setting will apply to them.
	IntervalSeconds int `yaml:"intervalSeconds" default:"0"`
	// The local IP address of the server that the agent exposes to which
	// collectd will send metrics.  This defaults to an arbitrary address in
	// the localhost subnet, but can be overridden if needed.
	WriteServerIPAddr string `yaml:"writeServerIPAddr" default:"127.9.8.7"`
	// The port of the agent's collectd metric sink server.  If set to zero
	// (the default) it will allow the OS to assign it a free port.
	WriteServerPort uint16 `yaml:"writeServerPort" default:"0"`
	// This is where the agent will write the collectd config files that it
	// manages.  If you have secrets in those files, consider setting this to a
	// path on a tmpfs mount.  The files in this directory should be considered
	// transient -- there is no value in editing them by hand.  If you want to
	// add your own collectd config, see the collectd/custom monitor.
	ConfigDir string `yaml:"configDir" default:"/var/run/signalfx-agent/collectd"`

	// The following are propagated from the top-level config
	BundleDir            string `yaml:"-"`
	HasGenericJMXMonitor bool   `yaml:"-"`
	// Assigned by manager, not by user
	InstanceName string `yaml:"-"`
	// A hack to allow custom collectd to easily specify a single monitorID via
	// query parameter
	WriteServerQuery string `yaml:"-"`
}

// Validate the collectd specific config
func (cc *CollectdConfig) Validate() error {
	if !validCollectdLogLevels.Has(cc.LogLevel) {
		return fmt.Errorf("invalid collectd log level %s, valid choices are %v",
			cc.LogLevel, validCollectdLogLevels)
	}

	return nil
}

// Hash calculates a unique hash value for this config struct
func (cc *CollectdConfig) Hash() uint64 {
	hash, err := hashstructure.Hash(cc, nil)
	if err != nil {
		log.WithError(err).Error("Could not get hash of CollectdConfig struct")
		return 0
	}
	return hash
}

// WriteServerURL is the local address served by the agent where collect should
// write datapoints
func (cc *CollectdConfig) WriteServerURL() string {
	return fmt.Sprintf("http://%s:%d/", cc.WriteServerIPAddr, cc.WriteServerPort)
}

// InstanceConfigDir is the directory underneath the ConfigDir that is specific
// to this collectd instance.
func (cc *CollectdConfig) InstanceConfigDir() string {
	return filepath.Join(cc.ConfigDir, cc.InstanceName)
}

// ConfigFilePath returns the path where collectd should render its main config
// file.
func (cc *CollectdConfig) ConfigFilePath() string {
	return filepath.Join(cc.InstanceConfigDir(), "collectd.conf")
}

// ManagedConfigDir returns the dir path where all monitor config should go.
func (cc *CollectdConfig) ManagedConfigDir() string {
	return filepath.Join(cc.InstanceConfigDir(), "managed_config")
}

// StoreConfig holds configuration related to config stores (e.g. filesystem,
// zookeeper, etc)
type StoreConfig struct {
	OtherConfig map[string]interface{} `yaml:",inline,omitempty" default:"{}"`
}

// ExtraConfig returns generic config as a map
func (sc *StoreConfig) ExtraConfig() map[string]interface{} {
	return sc.OtherConfig
}

var (
	// EnvReplacer replaces . and - with _
	EnvReplacer = strings.NewReplacer(".", "_", "-", "_")
)

// BundlePythonHomeEnvvar returns an envvar string that sets the PYTHONHOME envvar to
// the bundled Python runtime.  It is in a form that is ready to append to
// cmd.Env.
func BundlePythonHomeEnvvar() string {
	bundleDir := os.Getenv(constants.BundleDirEnvVar)
	if runtime.GOOS == "windows" {
		return "PYTHONHOME=" + filepath.Join(bundleDir, "python")
	}
	return "PYTHONHOME=" + bundleDir
}
